### DISCUSSION TITLE
Consider allowing to add/remove receivers from `Select` dynamically

### DISCUSSION THREAD

**Discussion by leandro-lucarella-frequenz** [frequenz-floss/frequenz-channels-python#24] [Ideas]
We have some use cases where we need to add/remove receivers to a `Select` inside the `.ready()` loop. For now we need to create a second loop where `Select` is created, then `break` the inside (`.ready()`) loop and re-generate the `Select` object, which is inconvenient from the POV of the code, and also makes the code harder to read, as the flow becomes more complicated.

Another **big issue** with this approach is if it's not done very carefully, we can lose messages from receivers if the `break` is done before all available messages from all receivers in the current loop are not handled. The `break` should always be the last in the loop.

This also applies to `Merge` and `MergeNamed`.

Example:

```python
        while True:
            receivers = await gen_component_receivers(
                self.component_infos, self.microgrid_client
            )


            # Create a task that will periodically send formula results.
            resend_formulas_task = asyncio.create_task(self.resend_formulas())
            select = Select(
                component_data_receiver=Merge(*receivers.values()),
                config_update_receiver=self._config_update_receiver,
            )
            while await select.ready():
                if msg := select.component_data_receiver:
                    # Update symbols and recalculate formulas.
                    updated_symbols = self.formula_calculator.update_symbol_values(
                        msg.inner
                    )
                    self.formula_calculator.compute(
                        updated_symbols, only_formula_names=set(self._outputs.keys())
                    )
                elif msg := select.config_update_receiver:
                    try:
                        await self._reinitialize(msg.inner, resend_formulas_task)
                    except Exception as err:  # pylint: disable=broad-except
                        logger.error(
                            "Reinitialization failed, because an error occurred "
                            "while applying the new configuration: %s",
                            err,
                        )
                    else:
                        # Break from the inner loop to regenerate component receivers
                        break
```

https://github.com/frequenz-floss/frequenz-sdk-python/blob/4ce090ce32f539014101f145e8f97fc67990fa9e/src/frequenz/sdk/data_ingestion/microgrid_data.py#L229-L260

Other use case came from here:

* https://github.com/frequenz-floss/frequenz-sdk-python/pull/32#discussion_r1005330301

**Comment by leandro-lucarella-frequenz** [comment]
This will probably be closed in favor of #29, as we don't want to add new ways to misuse `Select`, but I will leave it open until the docs are ready just in case.

**Comment by leandro-lucarella-frequenz** [comment]
We will not pursue this feature and improve documentation (#29) to make it clear that `Select` has a very specific and narrow use case and shouldn't be used if this kind of flexibility is needed.

**Comment by llucax** [comment]
The current proposal is to implement this externally, by adding support to add/remove channels to a `merge` as defined in:

* #116
